### PR TITLE
feat: dynamic fields

### PR DESCRIPTION
**Pull request by mdeljavan**

### CODE REVIEW

File: src/shared/forminator/DynamicFields/MultipleSelectQuestion.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  } = useQuestionContext();
+
+  const options = React.useMemo(() => {
+    return choices?.map((opt) => ({ value: opt, label: opt })) ?? [];
+  }, [choices]);
+
+  return (
+    <>
+      <SelectMultiAutoComplete filterSelectedOptions label={label} options={options} />
+      {helpText && (
```
Comment (mh-alahdadian) [inline src/shared/forminator/DynamicFields/MultipleSelectQuestion.tsx:25]: move helpText to textField

File: src/shared/forminator/__stories__/FiledRenderer.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+import { action } from '@storybook/addon-actions';
+import { storiesOf } from '@storybook/react';
+
+import DictInput from '../inputs/dict-input/DictInput';
+import SubmitButton from '../utils/SubmitButton';
+import { DynamicFields, Forminator, Question, QuestionType } from '..';
+
+const questions: QuestionType[] = [
```
Comment (mh-alahdadian) [inline src/shared/forminator/__stories__/FiledRenderer.stories.tsx:9]: why QuestionType ?

File: src/shared/forminator/index.ts [hunk 3]
```diff
@@ -13,3 +13,7 @@ export { default as SelectMultiAutoComplete } from './inputs/SelectMultiAutoComp
 export { default as ConstantInput } from './inputs/ConstantInput';
 export { default as FragmentPrompt } from './utils/FragmentPrompt';
 export { default as CheckboxInput } from './inputs/CheckboxInput';
+export { DynamicFields } from './DynamicFields/DynamicFields';
+export type { Question as QuestionType } from './DynamicFields/DynamicFields';
```
Comment (mh-alahdadian) [inline src/shared/forminator/index.ts:17]: use IQuestion or QuestionInterface

File: src/shared/project-output/AnswerOutput.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  questionId: string;
+  questionType: QuestionType;
+}
+
+type Props = React.PropsWithChildren<OwnProps>;
+
+export function AnswerOutput(props: Props) {
+  const { answers, questionId, questionType } = props;
+
+  const answer = answers.find((ans) => ans.questionId === questionId);
```
Comment (mh-alahdadian) [inline src/shared/project-output/AnswerOutput.tsx:17]: try move this logic out

File: src/shared/project-output/AnswerOutput.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+  const text = questionType === 'CHECKBOX_MULTIPLE' ? answer?.value?.replace(',', '، ') : answer?.value;
+
+  return <MultilineOutput value={text ?? null} enableTruncating />;
```
Comment (mh-alahdadian) [inline src/shared/project-output/AnswerOutput.tsx:21]: you might use this directly

**Review by mh-alahdadian — CHANGES_REQUESTED** [review]

File: src/core/domain/Question.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+export type Questions = RoundQuestions_selfReview | RoundQuestions_peerReview | RoundQuestions_managerReview;
+
+export interface Question {
+  readonly id: string;
+  readonly questionType: QuestionType;
+  readonly order: number;
+  readonly label: string;
+  readonly helpText: string | null;
+  readonly choices: ReadonlyArray<string> | null;
+  readonly privateAnswerToPeerReviewers: boolean;
```
Comment (mh-alahdadian) [inline src/core/domain/Question.ts:16]: remove these booleans if you don't use them

File: src/pages/manager-review-page/achievements/ManagerReviewAchievementsRatingGroup.tsx [hunk 7]
```diff
@@ -52,9 +53,7 @@ export const ManagerReviewAchievementsRatingGroup = React.memo(function ManagerR
   const peerAnswers = (answers: Answers) =>
     innerJoin((a, b) => a.questionId === b.id, answers, peerReviewProjectQuestions);
 
-  const filteredByRating = rating
-    ? comments.filter((comment) => comment.rating === rating)
-    : comments.filter((comment) => comment.rating === rating && peerAnswers(comment.answers).every(prop('value')));
+  const filteredByRating = comments.filter((comment) => comment.rating === rating);
```
Comment (mh-alahdadian) [inline src/pages/manager-review-page/achievements/ManagerReviewAchievementsRatingGroup.tsx:56]: what happened to rest of the logic ?

File: src/shared/forminator/utils/FragmentPrompt.ts [hunk 8]
```diff
@@ -22,6 +22,8 @@ function FragmentPrompt<VF, VS>(props: Props<VF, VS>) {
   const fragment = useFragmentContext<VF>();
   const value = useFragmentValue<VF>();
   const when = value !== undefined && !equal(value, props.value);
+  console.log(value, props.value, when);
```
Comment (mh-alahdadian) [inline src/shared/forminator/utils/FragmentPrompt.ts:25]: remove

File: src/shared/forminator/inputs/SelectMultiAutoComplete.tsx [hunk 9]
```diff
@@ -20,7 +20,7 @@ interface OwnProps<Suggestion extends BaseSuggestion = BaseSuggestion> {
   initialValue?: string[]; // id
   options: Array<Suggestion>;
   label: string;
-  textFieldOptions?: OutlinedTextFieldProps;
+  textFieldOptions?: Omit<OutlinedTextFieldProps, 'variant'>;
```
Comment (mh-alahdadian) [inline src/shared/forminator/inputs/SelectMultiAutoComplete.tsx:23]: why not TextFieldProps instead ?

File: src/shared/utils/questionsAnswersPair.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { Answers, Questions } from 'src/core/domain';
+
+export const getQuestionsAnswersPair = (questions: Questions, answers: Answers): [string, string][] => {
+  return [...questions].map((question) => {
```
Comment (mh-alahdadian) [inline src/shared/utils/questionsAnswersPair.ts:4]: questions.map

File: src/shared/utils/transformAnswers.ts [hunk 11]
```diff
@@ -13,18 +13,25 @@ interface Question {
 }
 
 export const transformAnswersToInput = (ans: Record<string, any>, questions: ReadonlyArray<Question>) => {
-  return Object.entries(ans).map(([key, value]) => {
+  return Object.entries(ans).reduce<Answer[]>((acc, [key, value]) => {
```
Comment (mh-alahdadian) [inline src/shared/utils/transformAnswers.ts:16]: use old version like this 
`Object.entries(ans).filter(value).map`

File: src/shared/utils/transformAnswers.ts [hunk 12]
```diff
@@ -13,18 +13,25 @@ interface Question {
...
 
-export const transformAnswersToFormData = (answers: Answers, questions: ReadonlyArray<Question>) => {
+export const transformAnswersToFormData = (answers: Answers, questions: Questions) => {
   return answers.reduce((acc, curr) => {
-    const question = questions.find((q) => q.id === curr.questionId);
+    if (!curr.value) {
+      return acc;
+    }
+
+    const question = [...questions].find((q) => q.id === curr.questionId);
```
Comment (mh-alahdadian) [inline src/shared/utils/transformAnswers.ts:34]: questions.find

**Review by mh-alahdadian — CHANGES_REQUESTED** [review]

File: src/pages/manager-review-page/achievements/ManagerReviewAchievementsRatingGroup.tsx [hunk 13]
```diff
@@ -40,11 +48,15 @@ export const ManagerReviewAchievementsRatingGroup = React.memo(function ManagerR
...
+
+  const peerAnswers = (answers: Answers) =>
+    innerJoin((a, b) => a.questionId === b.id, answers, peerReviewProjectQuestions);
 
   const filteredByRating = rating
     ? comments.filter((comment) => comment.rating === rating)
-    : comments.filter((comment) => comment.rating === rating && comment.text);
-  const filteredComments = filteredByRating.filter((comment) => comment.text);
+    : comments.filter((comment) => comment.rating === rating && peerAnswers(comment.answers).every(prop('value')));
+  const filteredComments = filteredByRating.filter((comment) => peerAnswers(comment.answers).every(prop('value')));
```
[the remark this answers is not in the document]
Comment (mdeljavan) [reply src/pages/manager-review-page/achievements/ManagerReviewAchievementsRatingGroup.tsx:57]: no. revert

**Review by mehrdad-e — APPROVED** [review]

**Pull request merged**